query/src/sql/plan_parser.rs

```diff
@@ -793,43 +778,27 @@ impl PlanParser {
                         .ctx
                         .get_catalog()
                         .get_table_function(&table_name, Some(table_args))?;
-                    meta_id = table_func.get_id();
-                    meta_version = table_func.get_table_info().ident.version;
                     table = table_func.as_table();
                 } else {
                     table = self.ctx.get_table(&db_name, &table_name)?;
-                    meta_id = table.get_id();
-                    meta_version = table.get_table_info().ident.version;
                 }
 
-                let scan = {
-                    let tbl_scan_info = TableScanInfo {
-                        table_id: meta_id,
-                        table_version: Some(meta_version),
-                        table_schema: &table.schema(),
-                    };
-                    PlanBuilder::scan(&db_name, tbl_scan_info, None, None)
-                        .and_then(|builder| builder.build())
-                };
+                // TODO(xp): is it possible to use get_cluster_table_io_context() here?
+                let io_ctx = self.ctx.get_single_node_table_io_context()?;
+                let io_ctx = Arc::new(io_ctx);
 
-                // TODO: Move ReadSourcePlan to SelectInterpreter
                 let partitions = self.ctx.get_settings().get_max_threads()? as usize;
-                scan.and_then(|scan| match scan {
-                    PlanNode::Scan(ref scan) => {
-                        // TODO(xp): is it possible to use get_cluster_table_io_context() here?
-
-                        let io_ctx = self.ctx.get_single_node_table_io_context()?;
-                        table
-                            .read_plan(
-                                Arc::new(io_ctx),
-                                Some(scan.push_downs.clone()),
-                                // TODO(xp): remove partitions, partitioning hint has been included in io_ctx.max_threads and io_ctx.query_nodes
-                                Some(partitions),
-                            )
-                            .map(PlanNode::ReadSource)
-                    }
-                    _unreachable_plan => panic!("Logical error: Cannot downcast to scan plan"),
-                })
+
+                // TODO: Move ReadSourcePlan to SelectInterpreter
+                let source_plan = table.read_plan(
```

Comment:
table.read_plan will get the statistics by push_down, if we give it a default Extras, it'll return full statistics.

A quick fix is to read_plan 2 times.

1. read_plan in parser to build source_plan with default extras.
2. read_plan in plan_scheduler to build really source_plan.

But read_plan seems to be very expensive.

Reply:
It's better than before, let's merge it firstly!